### PR TITLE
Add support for COPY IN protocol

### PR DESCRIPTION
**Pull request by flimzy**
This PR is still a bit rough.  It does not have tests for corner cases (most important: when client sends CopyFail), and I'm not entirely sure that the structure is as simple as it can be.

But I'd love to get some feedback on the approach.

Two main API changes:

1. `CopyIn` added to the `DataWriter` type.
2. The `Execute` method in the `PortalCache` interface needs an additional argument. I've implemented this in a backward-compatible way, though.

Still to do:

- [ ] Clean up commit history
- [ ] What error code should the server return when the client sends a `CopyFail`?
- [ ] Test for `CopyFail` from client
- [ ] Test for Text format
- [ ] Test for mixed column formats
- [ ] Should we provide any parsing of the copy stream?  IMO, probably not by default (I don't need to parse it for my own needs), but maybe as helper utilities, in future PRs, if needed by someone?

Note that this does _not_ implement `CopyOutResponse` or `CopyBothResponse`. I don't (yet) need either of these. But with the groundwork in this PR, they should both be easier to accomplish if/when needed.

### CODE REVIEW

**Review by jeroenrinzema — CHANGES_REQUESTED** [review]
Thanks for opening a PR! Left a few comments. Would like to explore if we could expose the copy implementation as a `io` interface.

File: writer.go [hunk 1]
```diff
@@ -79,6 +86,47 @@ func (writer *dataWriter) Row(values []any) error {
...
+	if writer.copy == nil {
+		return errors.New("DataCopyFn is nil; use PortalCacheCopy to execute CopyIn")
+	}
+	writer.client.Start(types.ServerCopyInResponse)
+	writer.client.AddByte(byte(format))
+	const n = 3
+	writer.client.AddInt16(n)
+	for i := 0; i < n; i++ {
+		writer.client.AddInt16(0)
+	}
```
[the remark this answers is not in the document]
Comment (flimzy) [reply writer.go]: Moved it to a separate method with proper godoc.

File: command.go [hunk 2]
```diff
@@ -130,101 +143,148 @@ func (srv *Server) handleMessageSizeExceeded(reader *buffer.Reader, writer *buff
...
-			return err
+// copyDataFn returns the default [CopyDataFn] function.
+func (srv *Server) copyDataFn(reader *buffer.Reader, writer *buffer.Writer) CopyDataFn {
+	return func(ctx context.Context) ([]byte, error) {
+		var results []byte
+		err := srv.commandLoop(ctx, reader, writer, srv.handleCopyInCommand(func(r []byte) { results = r }))
+		if err == errClientCopyDone {
+			err = io.EOF
 		}
+		srv.logger.Debug("srv.copyDataFn", slog.String("results", string(results)))
```
[the remark this answers is not in the document]
Comment (flimzy) [reply command.go:251]: Just some debugging I failed to clean up.

File: command.go [hunk 3]
```diff
@@ -130,101 +143,148 @@ func (srv *Server) handleMessageSizeExceeded(reader *buffer.Reader, writer *buff
...
+}
 
-		err = conn.Close()
-		if err != nil {
-			return err
+// copyDataFn returns the default [CopyDataFn] function.
+func (srv *Server) copyDataFn(reader *buffer.Reader, writer *buffer.Writer) CopyDataFn {
+	return func(ctx context.Context) ([]byte, error) {
+		var results []byte
+		err := srv.commandLoop(ctx, reader, writer, srv.handleCopyInCommand(func(r []byte) { results = r }))
```
[the remark this answers is not in the document]
Comment (flimzy) [reply command.go:247]: I had the same thought to turn it into a reader. So good call. Done.

[the remark this answers is not in the document]
Comment (flimzy) [reply command.go:247]: I agree this is a bit cumbersome. I'm not sure that an io.Writer or goroutine would help much, though.

We're not actually reading anything extra into memory as it is.  The `results`  value just gets the contents of `reader.Msg`, which is already completely in memory.

The reason for this awkward callback is to facilitate a common function signature for `handleCommand` and `handleCopyInCommand`, so that they can both be passed to `commandLoop` (formerly part of `consumeCommands`.

[the remark this answers is not in the document]
Comment (flimzy) [reply command.go]: I've refactored this bit now, to be quite a bit simpler, I think. I'm curious to hear your thoughts on how it looks now.

[the remark this answers is not in the document]
Comment (jeroenrinzema) [reply command.go]: Interesting approach, it is indeed a bit simpler. Got a few ideas I was thinking about which I would play with tonight to check if we could avoid having to copy the buffer.

[the remark this answers is not in the document]
Comment (jeroenrinzema) [reply command.go]: What if we pass the content length to the handler function. This way we could stream the incoming bytes to the returned io.Reader and terminate the io.Reader when reaching the total byte length. It might make sense to take over the buffered reader in case of a copy command in order to avoid the `ErrMessageSizeExceeded` error being thrown for larger payloads.

[the remark this answers is not in the document]
Comment (flimzy) [reply command.go]: Which content length are you referring to?  For the individual CopyData packets?  Or the overall length of the copy data?  I don't think the latter is communicated to the server (or even known by the client in many cases).  But maybe I'm overlooking something obvious.

[the remark this answers is not in the document]
Comment (jeroenrinzema) [reply command.go]: I am not able to test this at the moment, typed messages within the psql wire protocol have a specific length. We need to check whether the message length is send for the entire copydata message or whether the messages are split in chunks. The message length is fetched over here: https://github.com/jeroenrinzema/psql-wire/blob/main/pkg/buffer/reader.go#L125-L131

[the remark this answers is not in the document]
Comment (flimzy) [reply command.go]: I still don't quite follow.

A single COPY IN is typicall comprised of multile CopyData packets.

I don't think there's any way to know the overall length of the entire stream of data to be copied in, until the CopyDone packet is received.

[the remark this answers is not in the document]
Comment (jeroenrinzema) [reply command.go]: Apologies for the late response I have been sick over the past week. Good point, I missed that within the PSQL wire protocol implementation. The way how we could do it is as following:

```go
type copyDataReader struct {
	reader *buffer.Reader
	size   int
}

func (r *copyDataReader) Read(p []byte) (int, error) {
	if r.size == 0 {
		_, err := r.reader.ReadType()
		if err != nil {
			return -1, err
		}

		r.size, err = r.reader.ReadMsgSize()
		if err != nil {
			return -1, err
		}
	}

	n, err := io.ReadFull(r.reader.Buffer, p)
	r.size -= n
	return n, err
}
```

We need to expose a bit more internal logic within the buffered reader but this would allow us to bypass copying the buffer twice. The reader directly interacts with the wire protocol and passes the given buffer by the reader to `io.ReadFull`.  I have just opened a PR exposing the methods used in the example above: https://github.com/jeroenrinzema/psql-wire/pull/74.

[the remark this answers is not in the document]
Comment (jeroenrinzema) [reply command.go]: @flimzy I have implemented the changes discussed above and have included a simple binary column reader for the copy reader. Feel free to check it out!

**Comment by jeroenrinzema** [comment]
Much thanks @flimzy! I have implemented the last changes as discussed in the comments. This PR is ready to be merged!

**Review by jeroenrinzema — APPROVED** [review]
